### DISCUSSION TITLE
Behaviour of firejail is not consistent with whitelist, read-only rules. Works for some case, does not work for some cases. Would require some understanding for some queries!

### DISCUSSION THREAD

**Discussion by bodachaitanya** [netblue30/firejail#5489] [Q&A]
Hi Firejail team,

Could you please help in clarification for below queries, which I am seeing inconsistent behavior in my environment (on CentOS-7.9):

1. Suppose I have /etc/hosts as softlink to /data/runtime/hosts, something like below:
    ```
   /etc/hosts -> /data/runtime/hosts
   ```
   Having below rules was not allowing firejail to run:
   ```
   read-only /etc/hosts
   whitelist /data/runtime/hosts
   read-only /data/runtime/hosts
   ```
  **Context**: I have experienced this problem when I run ping in firejail with FQDN name. The DNS resolution is configured to go via non-default interface (which is achieved by preloading internal library which has modified connect, sendto, bind API's). Same is the case with /etc/resolv.conf, /etc/nsswitch.conf files, for this specific use-case. Firejail always comes out without executing them with above rules in place. When I remove whitelist (eg: in above case `whitelist /data/runtime/hosts`), this scenario works fine. 
Could anyone please explain the behavior of firejail here.

2. Suppose I have a situation something like this:
   - /var/lib -> /data  (/var/lib is a softlink to /data)
   - I have a `read-only /var/lib/runtime` rule in common profile.
   - I have another rule in application profile `whitelist /data/runtime/temp`, which is added before including common profile which has above read-only rule for same path, something like this in sequence:
   ```
   whitelist /data/runtime/temp
           :
   read-only /var/lib/runtime
   ```
   - I noticed firejail comes out, without any error.

**Context**:  I am trying to run tcpdump in firejail with above situation with filters having FQDN name (eg: host google.com), and tcpdump will create runtime files (along with snoop file) in /data/runtime/temp/ path.  I noticed firejail comes out without any error not executing as desired and the behaviour is inconsistent. Could you please explain this behavior.
    
3. I also have some issues with tools running in firejail with some rules configured (with whitelist and read-only) only in case of tools attempted to get DNS resolution from outside. 

For all the above queries, one thing is common. When network library calls are intercepted through the preloaded library, and with DNS context, the rules which were working before as mentioned in above examples, were started not working, and the behavior is not understable from firejail pov. Could you please help in here in understanding firejail behavior, what is causing the conflicts and how to analyse it.

Please let me know for any more information. Thanks

**Comment by bodachaitanya** [comment]
Hi @rusty-snake ,
I have checked the duplicate tickets mentioned in https://github.com/netblue30/firejail/issues/5490 (created for the same topic), and I understand /data is not supported for whitelist. 
With whitelist on /data/.. paths were actually not working. Whereas, blacklist and read-only for /data/.. paths seems to have no problem.
Also, I have firejail running fine with whitelist rules for /data/... paths before above changes.
Could you please help in understanding how is this internally handled in firejail?

**Reply by rusty-snake** [reply]
Since 0.9.70:

> The top  directory  can  be  **all  directories  in  /**  (except  /proc  and  /sys),  /sys/module, /run/user/$UID, $HOME and all directories in /usr.
